README.md
# rstml

[![crates.io page](https://img.shields.io/crates/v/rstml.svg)](https://crates.io/crates/rstml)
[![docs.rs page](https://docs.rs/rstml/badge.svg)](https://docs.rs/rstml/)
[![codecov](https://codecov.io/gh/rs-tml/rstml/branch/main/graph/badge.svg?token=2LMJ8YEV92)](https://codecov.io/gh/rs-tml/rs-tml)
![build](https://github.com/rs-tml/rstml/workflows/ci/badge.svg)
![license: MIT](https://img.shields.io/crates/l/rstml.svg)
Rust templating for XML-based formats (HTML, SVG, MathML) implemented on top of `proc-macro::TokenStream`s.
Similar to JSX but for Rust (commonly named RSX).The parsed result is a nested `Node` structure, similar to the browser DOM, where node name and value are syn expressions to support building proc macros.

The fork of original [syn-rsx](https://github.com/stoically/syn-rsx) repo.
It was created because of various reasons:
- The author of original `syn-rsx` retired, end left new features without attention https://github.com/stoically/syn-rsx/issues/49
- Syn v2 was released and `syn-rsx` need to be mooved to new version.
- The idea of [lossless parsing](https://github.com/stoically/syn-rsx/issues/53) was left unattended.
- [Unquoted text](https://github.com/stoically/syn-rsx/issues/2) feature should advance.
- Interest in recoverable parsing and better IDE support.


See [comparsion](/comparsion-with-syn-rsx.md) for more detail.

```rust
use std::convert::TryFrom;

use eyre::bail;
use quote::quote;
use rstml::{parse2, Node, NodeAttribute, NodeElement, NodeText};

// Create HTML `TokenStream`.
let tokens = quote! { <hello world>"hi"</hello> };

// Parse the tokens into a tree of `Node`s.
let nodes = parse2(tokens)?;

// Extract some specific nodes from the tree.
let Node::Element(element) = &nodes[0] else { bail!("element") };
let Node::Attribute(attribute) = &element.attributes[0] else { bail!("attribute") };
let Node::Text(text) = &element.children[0] else { bail!("text") };

// Work with the nodes.
assert_eq!(element.name.to_string(), "hello");
assert_eq!(attribute.key.to_string(), "world");
assert_eq!(String::try_from(&text.value)?, "hi");
```

You might want to check out the [html-to-string-macro example] as well.

## Features

- **Not opinionated**

  Every tag or attribute name is valid

  ```html
  <hello world />
  ```

- **Text nodes**


  ```html
  <div>"String literal"</div>
  ```

- **Unquoted text nodes**

  Unquoted text is supported with few limitations:
  - Only valid Rust TokenStream can be unquoted text (no single quote text is supported, no unclosed braces, etc.)
  - Unquoted text not always can save spaces. It uses [`Span::source_text`] and [`Span::join`] to retrive info about spaces, and it is not always available.
  - Quoted text near unquoted treated as diferent Node, end library user should decide whenever to preserve quotation.

  ```html
  <div> Some string that is valid rust token stream </div>
  ```

- **Node names separated by dash, colon or double colon**

  ```html
  <tag-name some:attribute-key="value" />
  <tag::name attribute::key="value" />
  ```

- **Node names with reserved keywords**

  ```html
  <input type="submit" />
  ```

- **Doctypes, Comments and Fragments**

  ```html
  <!DOCTYPE html>
  <!-- "comment" -->
  <></>
  ```

- **Braced blocks are parsed as arbitrary Rust code**

  ```html
  <{ let block = "in node name position"; } />
  <div>{ let block = "in node position"; }</div>
  <div { let block = "in attribute position"; } />
  <div key={ let block = "in attribute value position"; } />
  ```

- **Attribute values can be any valid syn expression without requiring braces**

  ```html
  <div key=some::value() />
  ```

- **Helpful error reporting out of the box**

  ```rust,no-run
  error: open tag has no corresponding close tag and is not self-closing
  --> examples/html-to-string-macro/tests/lib.rs:5:24
    |
  5 |     html_to_string! { <div> };
    |                        ^^^
  ```

- **Possibility to get the span for a whole node**

  This can be used to improve error reporting, e.g.

  ```rust,no-run
  error: Invalid element
  --> examples/src/main.rs:14:13
     |
  14 | /             <div>
  15 | |                 "invalid node for some consumer specific reason"
  16 | |             </div>
     | |__________________^
     
  ```

- **Recoverable parser**

  Can parse html with multiple mistakes.
  As result library user get array of errors that can be reported, and tree of nodes that was parsed.

  ```html
   <div hello={world.} /> <!-- dot after world is invalid syn expression -->
    <>
        <div>"1"</x> <!-- incorrect closed tag -->
        <div>"2"</div>
        <div>"3"</div>
        <div {"some-attribute-from-rust-block"}/>
    </>
  ```

  Using this feature one can write macro in IDE friendly way.
  This macro will work faster (because on invalid syntax it change output slightly, instead of removing it completely, so IDE can check diff quicly). And give completion (goto definition, and other semantic related feature) more often.

- **Customization**

  A `ParserConfig` to customize parsing behavior is available, so if you have
  slightly different requirements for parsing and it's not yet customizable
  feel free to open an issue or pull request to extend the configuration.

  One highlight with regards to customization is the `transform_block`
  configuration, which takes a closure that receives raw block content as
  `ParseStream` and lets you optionally convert it to a `TokenStream`. That makes it
  possible to have custom syntax in blocks. More details in [#9]

[`syn`]: https://github.com/dtolnay/syn
[`Span::join`]: https://doc.rust-lang.org/proc_macro/struct.Span.html#method.join
[`Span::source_text`]: https://doc.rust-lang.org/proc_macro/struct.Span.html#method.source_text
[`tokenstream`]: https://doc.rust-lang.org/proc_macro/struct.TokenStream.html
[html-to-string-macro example]: https://github.com/stoically/syn-rsx/tree/main/examples/html-to-string-macro
[#9]: https://github.com/stoically/syn-rsx/issues/9


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.10.5] - 2023-05-17

### Chore

- Fix docs build.

## [0.10.4] - 2023-05-17

### Chore

- Fix fmt add git-hook

## [0.10.3] - 2023-05-14

### Chore

- Remove dbg macro.

## [0.10.2] - 2023-05-14

### Documentation

- Fixup cargo docs

## [0.10.0] - 2023-05-13

### Chore

- Refactor node module, split types into submodules
- Update to syn 2.0
- Implement config passing, and allow to emit more than one error from macro expansion.
- Make Node types clonable.
- Fix recoverable parser. Now try to recover if any sequence is incorrect.
- Bump syn_derive to 0.1.6 to avoid nightly
- Start a new version history

### Documentation

- Add comparsion with syn-rsx

### Features

- Remove attributes from node list, and make type guaranties that attribute will be stored only in Element
- Improve tag close handling.
- Allow parsing of invalid rust code blocks
- Add support of unqoted text.

### Fix

- Refactor ToTokens implementation and node structure to contain all source tokens.
- Re-implement tree flattening.
- Reimplement transform_block and fix tests
- Attribute value parsing span
- Updated example, bench, made fragment parsing recoverable.
- Make public keyed attribute fields.
- RawText to_source implementation on stable

### Refactor

- Move node.rs to separate module.
- Remove Display implementation for most of node types.
- Simlify Parse and ToToken impl, by using syn_derive.
- Refactor: remove NodeValueExpr type
NodeValueExpr was replaced by syn::Expr in all places where expr is needed.
For places where {code} is expected NodeBlock was used.

- Cleanup of parsing code.
- Refactor recoverable parser.
- Refactor: Remove tls context at all.
Use RecoverableContext instead.


### Testing

- TRYBUILD=overwrite in ui tests


## [unreleased]

### Documentation

- Mixed name punctuation and node span ([#44](https://github.com/stoically/syn-rsx/issues/44))
- Missing module documentation ([#45](https://github.com/stoically/syn-rsx/issues/45))
- Fix formatting ([#46](https://github.com/stoically/syn-rsx/issues/46))
- Fix code block ([#47](https://github.com/stoically/syn-rsx/issues/47))

### Features

- Track the source span for all nodes ([#42](https://github.com/stoically/syn-rsx/issues/42))

## [0.9.0] - 2022-11-10

### Documentation

- Remove outdated node link ([#36](https://github.com/stoically/syn-rsx/issues/36))
- Hint regarding braced blocks ([#38](https://github.com/stoically/syn-rsx/issues/38))
- Fix typo ([#39](https://github.com/stoically/syn-rsx/issues/39))

## [0.9.0-beta.2] - 2022-11-06

### Features

- [**breaking**] Replace Colon and Dash with a merged variant ([#34](https://github.com/stoically/syn-rsx/issues/34))

## [0.9.0-beta.1] - 2022-11-03

### Chore

- Add rustfmt.toml

### Documentation

- Update README
- Update example
- Update html macro docs
- Add link to example
- Fix node links
- Improve parser docs
- Remove TODO
- Fix attribute value example ([#28](https://github.com/stoically/syn-rsx/issues/28))
- Fix blocks example ([#29](https://github.com/stoically/syn-rsx/issues/29))
- Fix typo ([#30](https://github.com/stoically/syn-rsx/issues/30))

### Features

- [**breaking**] Make path_to_string private

### Other

- Move flat tree converter back into node parser
- Switch fmt to nightly toolchain

### Refactor

- [**breaking**] Drop `NodeName::span` method
- Pass block_transform a forked stream
- Move flat tree converter to node method
- Replace extrude with let-else ([#31](https://github.com/stoically/syn-rsx/issues/31))

## [0.9.0-alpha.1] - 2022-10-21

### Chore

- Add Cargo.lock to .gitignore
- Add git-cliff configuration
- Add CHANGELOG
- Use the actual html-to-string-macro crate as example
- Update README badges

### Refactor

- Move config into dedicated module
- [**breaking**] Switch `Node` to enum-style ([#23](https://github.com/stoically/syn-rsx/issues/23))

## [0.8.1] - 2022-06-26

### Chore

- Clippy
- Clippy
- Remove Cargo.lock
- Bump dependencies

### Documentation

- Update README

## [0.8.0] - 2021-02-17

### Documentation

- Fix and sync
- Typo

### Features

- Doctypes, comments and fragments
- Value_as_string support for `ExprPath`

### Other

- Should be value not name

### Refactor

- Remove unnecessary `Result`

### Testing

- More reserved keywords tests

## [0.8.0-beta.2] - 2021-01-04

### Documentation

- Sync lib with readme

### Features

- [**breaking**] Block in node name position ([#11](https://github.com/stoically/syn-rsx/issues/11))

### Other

- Tarpaulin and codecov

## [0.8.0-beta.1] - 2021-01-03

### Documentation

- Node

### Features

- Properly handle empty elements
- [**breaking**] Transform_block callback ([#9](https://github.com/stoically/syn-rsx/issues/9))
- [**breaking**] Doctype ([#6](https://github.com/stoically/syn-rsx/issues/6))
- [**breaking**] Html comments ([#7](https://github.com/stoically/syn-rsx/issues/7))
- [**breaking**] Fragments ([#8](https://github.com/stoically/syn-rsx/issues/8))

### Other

- Formatting
- Bump criterion

### Refactor

- Cleanup

## [0.7.3] - 2020-10-30

### Documentation

- Rephrase misleading unquoted text hint
- Update node description
- Update NodeName description

### Features

- Value_as_block method for nodes
- Implement ToString for NodeName
- Support blocks in html-to-string-macro
- Implement ToTokens for NodeName

### Other

- Only count top level nodes in case of flat_tree
- Parse2 with criterion
- More test tokens
- Update

### Performance

- More peeking and better block parsing ([#5](https://github.com/stoically/syn-rsx/issues/5))
- Use `node_name_punctuated_ident` to parse name path

### Refactor

- Better error reporting
- Rename test file
- Switch impl ToString on Node to impl Display
- Merge text and block handling

## [0.7.2] - 2020-09-10

### Documentation

- Error reporting

### Features

- Expose span fn on NodeName as well

### Refactor

- Better error messages

## [0.7.1] - 2020-09-09


src/lib.rs
//! Rust templating for XML-based formats (HTML, SVG, MathML) implemented on top
//! of [`proc-macro::TokenStream`]s. Similar to JSX but for Rust (commonly named
//! RSX). The parsed result is a nested [`Node`] structure, similar to the
//! browser DOM, where node name and value are syn expressions to support
//! building proc macros.
//!
//! ```rust
//! # fn main() -> eyre::Result<()> {
//! use std::convert::TryFrom;
//!
//! use eyre::bail;
//! use quote::quote;
//! use rstml::{
//!     node::{Node, NodeAttribute, NodeElement, NodeText},
//!     parse2,
//! };
//!
//! // Create HTML `TokenStream`.
//! let tokens = quote! { <hello world>"hi"</hello> };
//!
//! // Parse the tokens into a tree of `Node`s.
//! let nodes = parse2(tokens)?;
//!
//! // Extract some specific nodes from the tree.
//! let Node::Element(element) = &nodes[0] else { bail!("element") };
//! let NodeAttribute::Attribute(attribute) = &element.attributes()[0] else { bail!("attribute") };
//! let Node::Text(text) = &element.children[0] else { bail!("text") };
//!
//! // Work with the nodes.
//! assert_eq!(element.name().to_string(), "hello");
//! assert_eq!(attribute.key.to_string(), "world");
//! assert_eq!(text.value_string(), "hi");
//! # Ok(())
//! # }
//! ```
//!
//! You might want to check out the [html-to-string-macro example] as well.
//!
//! ## Features
//!
//! - **Not opinionated**
//!
//!   Every tag or attribute name is valid
//!
//!   ```rust
//!   # use quote::quote;
//!   # use rstml::parse2;
//!   # parse2(quote! {
//!   <hello world />
//!   # }).unwrap();
//!   ```
//!
//! - **Text nodes**
//!
//!
//!   ```rust
//!   # use quote::quote;
//!   # use rstml::parse2;
//!   # parse2(quote! {
//!   <div>"String literal"</div>
//!   # }).unwrap();
//!   ```
//!
//!
//! - **Unquoted text nodes**
//!
//! Unquoted text is supported with few limitations:
//! - Only valid Rust TokenStream can be unquoted text (no single quote text is
//!   supported, no unclosed braces, etc.)
//! - Unquoted text not always can save spaces. It uses [`Span::source_text`]
//!   and [`Span::join`] to retrive info about spaces, and it is not always
//!   available.
//! - Quoted text near unquoted treated as diferent Node, end library user
//!   should decide whenever to preserve quotation.
//!
//! ```rust
//! 
//!   # use quote::quote;
//!   # use rstml::parse2;
//!   # parse2(quote! {
//!    <div> Some string that is valid rust token stream </div>
//!   # }).unwrap();
//! ```
//!
//! - **Node names separated by dash, colon or double colon**
//!
//!   ```rust
//!   # use quote::quote;
//!   # use rstml::parse2;
//!   # parse2(quote! {
//!   <tag-name some:attribute-key="value" />
//!   <tag::name attribute::key="value" />
//!   # }).unwrap();
//!   ```
//!
//! - **Node names with reserved keywords**
//!
//!   ```rust
//!   # use quote::quote;
//!   # use rstml::parse2;
//!   # parse2(quote! {
//!   <input type="submit" />
//!   # }).unwrap();
//!   ```
//!
//! - **Doctypes, Comments and Fragments**
//!
//!   ```rust
//!   # use quote::quote;
//!   # use rstml::parse2;
//!   # parse2(quote! {
//!   <!DOCTYPE html>
//!   <!-- "comment" -->
//!   <></>
//!   # }).unwrap();
//!   ```
//!
//! - **Braced blocks are parsed as arbitrary Rust code**
//!
//!   ```rust
//!   # use quote::quote;
//!   # use rstml::parse2;
//!   # parse2(quote! {
//!   <{ let block = "in node name position"; } />
//!   <div>{ let block = "in node position"; }</div>
//!   <div { let block = "in attribute position"; } />
//!   <div key={ let block = "in attribute value position"; } />
//!   # }).unwrap();
//!   ```
//!
//! - **Attribute values can be any valid syn expression without requiring
//!   braces**
//!
//!   ```rust
//!   # use quote::quote;
//!   # use rstml::parse2;
//!   # parse2(quote! {
//!   <div key=some::value() />
//!   # }).unwrap();
//!   ```
//!
//! - **Helpful error reporting out of the box**
//!
//!   ```no_build
//!   error: open tag has no corresponding close tag and is not self-closing
//!   --> examples/html-to-string-macro/tests/lib.rs:5:24
//!     |
//!   5 |     html_to_string! { <div> };
//!     |                        ^^^
//!   ```
//!
//! - **Possibility to get the span for a whole node**
//!
//!  This can be used to improve error reporting, e.g.
//!
//!  ```no_build
//!  error: Invalid element
//!  --> examples/src/main.rs:14:13
//!     |
//!  14 | /             <div>
//!  15 | |                 "invalid node for some consumer specific reason"
//!  16 | |             </div>
//!     | |__________________^
//!  ```
//!
//! - **Recoverable parser**
//!
//! Can parse html with multiple mistakes.
//! As result library user get array of errors that can be reported, and tree of
//! nodes that was parsed.
//!
//! ```rust
//!   # use quote::quote;
//!   # use rstml::{Parser, ParserConfig};
//!   # Parser::new(ParserConfig::default()).parse_recoverable(quote! {
//!  <div hello={world.} /> <!-- dot after world is invalid syn expression -->
//!   <>
//!       <div>"1"</x> <!-- incorrect closed tag -->
//!       <div>"2"</div>
//!       <div>"3"</div>
//!       <div {"some-attribute-from-rust-block"}/>
//!   </>
//!   # });
//! ```
//!
//! Using this feature one can write macro in IDE friendly way.
//! This macro will work faster (because on invalid syntax it change output
//! slightly, instead of removing it completely, so IDE can check diff quicly).
//! And give completion (goto definition, and other semantic related feature)
//! more often.
//!
//! - **Customization**
//!
//!   A [`ParserConfig`] to customize parsing behavior is available, so if you
//! have   slightly different requirements for parsing and it's not yet
//! customizable   feel free to open an issue or pull request to extend the
//! configuration.
//!
//!   One highlight with regards to customization is the [`transform_block`]
//!   configuration, which takes a closure that receives raw block content as
//!   `ParseStream` and lets you optionally convert it to a `TokenStream`. That
//! makes it   possible to have custom syntax in blocks. More details in [#9]
//!
//!
//! [`syn`]: /syn
//! [`TokenStream`]: https://doc.rust-lang.org/proc_macro/struct.TokenStream.html
//! [`Node`]: enum.Node.html
//! [`Span::join`]: https://doc.rust-lang.org/proc_macro/struct.Span.html#method.join
//! [`Span::source_text`]: https://doc.rust-lang.org/proc_macro/struct.Span.html#method.source_text
//! [`ParserConfig`]: struct.ParserConfig.html
//! [mod style path]: https://docs.rs/syn/1.0.40/syn/struct.Path.html#method.parse_mod_style
//! [unquoted text is planned]: https://github.com/stoically/syn-rsx/issues/2
//! [`transform_block`]: struct.ParserConfig.html#method.transform_block
//! [#9]: https://github.com/stoically/syn-rsx/issues/9
//! [html-to-string-macro example]: https://github.com/stoically/syn-rsx/tree/main/examples/html-to-string-macro

extern crate proc_macro;

use syn::Result;

mod config;
mod error;
pub mod node;
mod parser;
pub use config::ParserConfig;
pub use error::Error;
pub use node::atoms;
use node::Node;
pub use parser::{recoverable, recoverable::ParsingResult, Parser};

/// Parse the given [`proc-macro::TokenStream`] into a [`Node`] tree.
///
/// [`proc-macro::TokenStream`]: https://doc.rust-lang.org/proc_macro/struct.TokenStream.html
/// [`Node`]: struct.Node.html
pub fn parse(tokens: proc_macro::TokenStream) -> Result<Vec<Node>> {
    Parser::new(ParserConfig::default()).parse_simple(tokens)
}

/// Parse the given [`proc-macro::TokenStream`] into a [`Node`] tree with custom
/// [`ParserConfig`].
///
/// [`proc-macro::TokenStream`]: https://doc.rust-lang.org/proc_macro/struct.TokenStream.html
/// [`Node`]: struct.Node.html
/// [`ParserConfig`]: struct.ParserConfig.html
#[deprecated(since = "0.10.2", note = "use rstml::Parser::parse_simple instead")]
pub fn parse_with_config(
    tokens: proc_macro::TokenStream,
    config: ParserConfig,
) -> Result<Vec<Node>> {
    Parser::new(config).parse_simple(tokens)
}
/// Parse the given [`proc-macro2::TokenStream`] into a [`Node`] tree.
///
/// [`proc-macro2::TokenStream`]: https://docs.rs/proc-macro2/latest/proc_macro2/struct.TokenStream.html
/// [`Node`]: struct.Node.html
pub fn parse2(tokens: proc_macro2::TokenStream) -> Result<Vec<Node>> {
    Parser::new(ParserConfig::default()).parse_simple(tokens)
}

/// Parse the given [`proc-macro2::TokenStream`] into a [`Node`] tree with
/// custom [`ParserConfig`].
///
/// [`proc-macro2::TokenStream`]: https://docs.rs/proc-macro2/latest/proc_macro2/struct.TokenStream.html
/// [`Node`]: struct.Node.html
/// [`ParserConfig`]: struct.ParserConfig.html
#[deprecated(since = "0.10.2", note = "use rstml::Parser::parse_simple instead")]
pub fn parse2_with_config(
    tokens: proc_macro2::TokenStream,
    config: ParserConfig,
) -> Result<Vec<Node>> {
    Parser::new(config).parse_simple(tokens)
}
